cql3/expr/prepare_expr.cc

```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
     return ::make_shared<assignment_testable_expression>(std::move(e));
 }
 
+// Finds column_defintion for given column name in the schema.
+// Second argument is a relation to print in the error message in case of failure.
+static const column_definition* resolve_column(const unresolved_identifier& col_ident, const schema& schema, const binary_operator& printable_relation) {
+    ::shared_ptr<column_identifier> id = col_ident.ident->prepare_column_identifier(schema);
+    const column_definition* def = get_column_definition(schema, *id);
+    if (!def || def->is_hidden_from_cql()) {
+        throw exceptions::unrecognized_entity_exception(*id, fmt::format("{}", printable_relation));
+    }
+    return def;
+}
+
+// Resolves columns on LHS of binary_operator and prepares the index in case of a subscripted column.
+// Last argument is a relation to print in the error message in case of failure.
+static expression prepare_binop_lhs(const expression& lhs, data_dictionary::database db, const schema& schema, const binary_operator& printable_relation) {
+    return expr::visit(overloaded_functor{
+        [&](const unresolved_identifier& unin) -> expression {
+            return column_value(resolve_column(unin, schema, printable_relation));
+        },
+        [](const column_value& cv) -> expression { return cv; },
+        [&](const tuple_constructor& tc) -> expression {
```

Comment:
Can this be part of the general prepare_expression()? Instead of just for binop lhs.

Reply:
It can't because we don't have a receiver for lhs

Reply:
right, we need to make the receiver optional and a two-phase type inference pass.